rcore/rdb.c

```diff
@@ -517,6 +529,49 @@ int rdb_insert(const struct rdb_database *db, uint8_t *key, uint16_t key_size, u
     return Blob_Success;
 }
 
+int rdb_update(const struct rdb_database *db, const uint8_t *key, const uint16_t key_size, const uint8_t *data, const size_t data_size)
+{
+    struct rdb_iter it;
+    rdb_select_result_list head;
+    assert(db);
+    list_init_head(&head);
+    int rv = rdb_iter_start(db, &it);
+    if (!rv) {
+        return Blob_InvalidDatabaseID;
+    }
+    
+    struct rdb_selector selectors[] = {
+        { RDB_SELECTOR_OFFSET_KEY, key_size, RDB_OP_EQ, key },
+        { }
+    };
+    int n = rdb_select(&it, &head, selectors);
+
+    if (!n) {
+        return Blob_KeyDoesNotExist;
+    }
+    
+    struct rdb_select_result *res;
+    rdb_select_result_foreach(res, &head) {
+        // is the data size <=
+        if (data_size != res->it.data_len) {
+            if (fs_write(&res->it.fd, data, data_size) < data_size) {
```

Comment:
This only works if the data is a subset (i.e., `old_data & new_data == new_data`) of the old data, right?

My plan in general was a generalized update mechanism on top of `insert` that always did a replace on an old key (I think there is a way to safely do it atomically), and then a `rdb_amend` op that works like this branch does, and demands that the old size be equal to the new size.

Reply:
that sounds good. One thing I noticed was an app continually writing to the same storage rdb. The data size was always the same, so I figured we just update the value save thrashing flash.
If the data size is not the same, we do the delete insert dance.
(also just noticed the comment doesn't match the logic any more)
I welcome our atomic update overlord

Reply:
It has just occured to me what you mean by this. Updating the flash in place isn't gonna work is it